### PR TITLE
[DI] leverage Contracts\Service

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Embedding #28206 for now.
This deprecates some generic interfaces from the DI component in favor of the similar ones proposed for `Contracts\Service`.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/ServiceLocator.php [hunk 1]
```diff
@@ -11,62 +11,37 @@
...
-    private $container;
-
-    /**
-     * @param callable[] $factories
-     */
-    public function __construct(array $factories)
-    {
-        $this->factories = $factories;
+    use ServiceLocatorTrait {
+        get as doGet;
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ServiceLocator.php]: does this make `doGet` public or no ?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ServiceLocator.php]: nope, visibility changes must be explicit.

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ServiceLocator.php]: ok, got it, fixed :)

**Comment by fabpot** [comment]
Should be rebased now

File: src/Symfony/Component/DependencyInjection/ResettableContainerInterface.php [hunk 2]
```diff
@@ -19,6 +19,8 @@
  * not needed anymore.
  *
  * @author Christophe Coevoet <stof@notk.org>
+ *
+ * @deprecated since Symfony 4.2, use `ResetInterface` instead.
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/ResettableContainerInterface.php]: should be the FQDN here IMHO

**Review by fabpot — APPROVED** [review]
minor comment

File: src/Symfony/Component/DependencyInjection/ResettableContainerInterface.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ResettableContainerInterface.php]: updated (and added to the CHANGELOG file)

**Review by stof — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**